src/org/rascalmpl/interpreter/utils/JavaCompiler.java

```diff
@@ -204,60 +209,157 @@ public synchronized Class<T> compile(final String qualifiedClassName,
+
+      // Write the bytes to their respective file URIs.
+      // Notice that there are possible many more .class files then there were java source files,
+      // because of nested and anonymous classes in their source code.
+      for (Entry<JavaFileObjectImpl, byte[]> e : javaFileManager.getAllClassBytes().entrySet()) {
+         var output = URIUtil.getChildLocation(target, e.getKey().toUri().getPath().replaceAll("\\.", "/"));
+         output = URIUtil.changeExtension(output, "class");
+
+         // this is where we implement support for any target scheme
+         try (OutputStream out = reg.getOutputStream(output, false)) {
+            out.write(e.getValue());
+         }
+         catch (IOException x) {
+            throw new JavaCompilerException(java.util.Set.of(e.getKey().toUri().getPath()), x, diagnostics);
+         }
+      }
+
+      return;
+   }
+
+   private List<JavaFileObject> registerSourceFiles(ISet classes) throws URISyntaxException {
+      List<JavaFileObject> sources = new ArrayList<JavaFileObject>();
+      
+      for (IValue entry : classes) {
+         ISourceLocation sloc = (ISourceLocation) ((ITuple) entry).get(0);
+         String qname = ((IString) ((ITuple) entry).get(1)).getValue();
+         CharSequence javaSource = ((IString) ((ITuple) entry).get(2)).getValue();
+
+         if (javaSource != null) {
+            final JavaFileObjectImpl source = new JavaFileObjectImpl(sloc, qname, javaSource);
+            sources.add(source);
+
+            int dotPos = qname.lastIndexOf('.');
+            String className = dotPos == -1 ? qname : qname.substring(dotPos + 1);
+            String packageName = dotPos == -1 ? "" : qname.substring(0, dotPos);
+
+            // Store the source file in the FileManager via package/class
+            // name.
+            // For source files, we add a .java extension
+            javaFileManager.putFileForInput(StandardLocation.SOURCE_PATH, packageName, className + JAVA_EXTENSION, source);
+         }
+      }
+
+      return sources;
+   }
+
+   private List<JavaFileObject> registerSourceFiles(final Map<String, CharSequence> classes) throws URISyntaxException {
+      List<JavaFileObject> sources = new ArrayList<JavaFileObject>();
+      
+      for (Entry<String, CharSequence> entry : classes.entrySet()) {
+         String qualifiedClassName = entry.getKey();
+         CharSequence javaSource = entry.getValue();
+         if (javaSource != null) {
+            int dotPos = qualifiedClassName.lastIndexOf('.');
+            String className = dotPos == -1 ? qualifiedClassName : qualifiedClassName.substring(dotPos + 1);
+            String packageName = dotPos == -1 ? "" : qualifiedClassName.substring(0, dotPos);
+            
+            JavaFileObjectImpl source = new JavaFileObjectImpl(className, javaSource);
+            sources.add(source);
+            javaFileManager.putFileForInput(StandardLocation.SOURCE_PATH, packageName, className + JAVA_EXTENSION, source);
```

Comment:
There is some duplication here with the previous function, can we factor that out?